lib/swaps/Swaps.ts

```diff
@@ -141,218 +142,284 @@ class Swaps extends EventEmitter {
-        responseBody = {
-          r_preimage,
-          dealId,
-        };
+        deal.r_preimage = Buffer.from(response.getPaymentPreimage_asB64(), 'base64').toString('hex');
+        this.setDealStatus(deal, SwapDealStatus.AmountReceived);
+        return deal.r_preimage;
       } catch (err) {
-        this.logger.error(`Got exception from sendPaymentSync ${JSON.stringify(request.toObject())}`, err);
-        this.emit('swap.failed', deal);
+        this.logger.error('Got exception from sendPaymentSync: ' + ' ' + JSON.stringify(request.toObject()) + err.message);
+        this.setDealState(deal, SwapDealState.Error, err.message);
+        return 'Got exception from sendPaymentSync' + err.message;
       }
     } else {
-      // this is not a valid deal
-      this.logger.warn(`received swap request for unrecognized deal ${dealId}`);
-      // TODO: respond with a failure message, possibly penalize peer
+      // If we are here we are the taker
+      this.logger.debug('Executing taker code');
+      this.setDealStatus(deal, SwapDealStatus.AmountReceived);
+      return deal.r_preimage;
     }
 
-    const responsePacket = new packets.SwapResponsePacket(responseBody, requestPacket.header.id);
+  }
 
-    this.logger.debug('Sending swap response to peer: ' + JSON.stringify(responseBody));
+  private setDealState = (deal: SwapDeal, newState: SwapDealState, newStateReason: string): void => {
+    deal.state = newState;
+    deal.stateReason = newStateReason;
+    this.emit('swap.failed', deal);
+  }
 
-    peer.sendPacket(responsePacket);
+  private setDealStatus = (deal: SwapDeal, newStatus: SwapDealStatus): void => {
+    assert(deal.state === SwapDealState.Normal, 'deal state is Error. Can not change deal status');
+    switch (newStatus) {
+      case SwapDealStatus.SwapCreated:
+        assert(false, 'can not set deal status to SwapCreated.');
+        break;
+      case SwapDealStatus.SwapRequested:
+        assert(deal.myRole === SwapDealRole.Taker, 'SwapRequested can only be set by the taker');
+        assert(deal.status === SwapDealStatus.SwapCreated, 'SwapRequested can be only be set after SwapCreated');
+        break;
+      case SwapDealStatus.SwapAgreed:
+        assert(deal.myRole === SwapDealRole.Maker, 'SwapAgreed can only be set by the maker');
+        assert(deal.status === SwapDealStatus.SwapCreated, 'SwapAgreed can be only be set after SwapCreated');
+        break;
+      case SwapDealStatus.AmountSent:
+        assert(deal.myRole === SwapDealRole.Taker && deal.status === SwapDealStatus.SwapRequested ||
+          deal.myRole === SwapDealRole.Maker && deal.status === SwapDealStatus.SwapAgreed,
+            'AmountSent can only be set after SwapRequested (taker) or SwapAgreed (maker)');
+        break;
+      case SwapDealStatus.AmountReceived:
+        assert(deal.status === SwapDealStatus.AmountSent, 'AmountReceived can be only be set after AmountSent');
+        break;
+      case SwapDealStatus.SwapDone:
```

Comment:
I'd prefer a status called `SwapComplete` - to match the packet name - but it can stay `SwapDone` if you prefer that.

Reply:
change to `swapCompleted`